### ISSUE TITLE
Go TypeSwitch

### ISSUE THREAD

**Issue by PXshadow** [go2hx/go2hx#8]
Currently can not use Abstracts as values for the "is" operator, blocked by needed compiler changes. https://github.com/HaxeFoundation/haxe/issues/2976

```haxe
 x = Go.assert(((1 : Byte)));
if ((x is UInt8)) { } else {
    panic("byte != uint8");
};
x = Go.assert(((2 : UInt8)));
if ((x is Byte)) { } else {
    panic("uint8 != byte");
};
var rune32 = false;
x = Go.assert(((3 : Rune)));
if ((x is Int)) { } else if ((x is Int32)) {
    rune32 = true;
} else {
    panic("rune != int and rune != int32");
};

**Comment by elliott5** [comment]
My prototype code for handling this issue is based on the following ideas:
1) keep the abstracts to make normal working fast for Go types
2) when converting that abstract to a Go 'interface{}' type (modelled as Haxe 'Any' type) place that abstract value into a Haxe Class for that purpose
3) all Go Classes must implement the GoInterface Haxe interface spec

[prototype](https://gist.github.com/elliott5/dd7e29e7d9d48f79a5e5fd75dd7f1e2a)

Interfaces and reflection are fundamental to Go, so we can't wait for a compiler fix - life's too short!

**Comment by elliott5** [comment]
While the abstracts for each emulated Go numeric data type will be different depending on which Haxe target is being compiled for, the Haxe Classes used to simulate Go interface{} entries will be the same. These Classes could therefore be auto-generated in some way. In TardisGo I wrote a Go function to spit them all out, I'm guessing there are probably better ways to do it in Haxe! ;)

**Comment by elliott5** [comment]
I just tried my prototype code on the cpp target and discovered that Haxe `Int` is `Int64` on that platform, which is entirely reasonable of course, as Int should be the default size for the target platform. Exactly the same logic applies for the `int`, `uint`, and `uintptr` types in Go.

**Comment by PXshadow** [comment]
I think instead of creating indvidual classes for each basic type, we could instead try to follow TinyGo's method of "Interfaces are always represented as a {typecode, value} pair"

**Comment by elliott5** [comment]
A {typecode, value} pair approach would certainly work, but would require a central registry of all the typecodes used in a particular compile. Since, if the project is successful, the Haxe generated by go2hx will be compiled into other people's Haxe programs, a central registry of typecodes will be difficult to maintain.

**Comment by PXshadow** [comment]
That is true, I don't however, see how it's possible to not have some kind of central registry with reflect (be that with a code/enum/or a switch for what type to class type to turn into). In refrence to this and to tie things together: https://github.com/PXshadow/go2hx/issues/11#issuecomment-779425358 . So the idea is to have each one of the generated Classes have fields for reflect to use? Wouldn't that add substantional overhead both for typing and required fields for a data type? Please correct me If I'm not understanding.  Also I'm opening up a new issue, to discuss whether to use typedefs, or classes here:  https://github.com/PXshadow/go2hx/issues/13 . I agree though it's probably best to go back to the initial idea to write reflect by hand, I'll use go2hx to generate the structs/fields and then fill them in using the functionality of Haxe's reflect class.

**Comment by elliott5** [comment]
The problem is that every unique Go type, even those that will be written by end-users in ten years time, need to have a unique typecode to index into the type information somewhere centrally.

So
```
type MyNewType struct {A, B int}
```
will need its own unique typecode.

I imagine the TinyGo compiler allocates these unique typecodes for each complier run. 

The only issue is finding a mechanism to link a Go type in an `interface{}` to the necessary type information for reflection.

**Comment by PXshadow** [comment]
Oh I see the issue, yeah I need to run some experiments and see how non basic types work when switching to Any/Dynamic, You're right with needing a unique typecode for each user type being an issue, and I'll look for a solution that doesn't require that, and only needs the typecode for basic types.

**Comment by PXshadow** [comment]
Current idea for implementation is abstract types can use a typeCode or typeName function, I think typeName makes more sense for readability, and the rest can use the is Operator.

**Comment by elliott5** [comment]
Determining the type of some datatype passed as an `interface{}` is central to how Go works.

I suggest resolving type identity by using fully-qualified (non-aliased) typeNames. This is slower than comparing typeCodes, but it does not require a central registry of typeCodes, important for providing go2hx generated code as Haxe libraries.

Clearly for go2hx generated Haxe classes, providing typeNames are no problem. 

Likewise for basic Haxe types (`Float`, `Int`, `String` ... others?) it is easy to generate the right Go typeName.

For abstract types this is more difficult, see the code below:
```
package command_line_arguments;

import stdgo.fmt.Fmt;
import stdgo.GoString;

class Cl {
	public function new() {}

	public function typeName():String {
		return "Cl";
	}
}

abstract Uint8(Int) {
	inline public function new(i:Int) {
		this = i;
	}

	inline public function typeName():String {
		return "uint8";
	}
}

private function main():Void {
	var i = 42;
	printType(i);
	var c = new Cl();
	printType(c);
	var ui8 = new Uint8(9);
	printType(ui8);
	stdgo.fmt.Fmt.println(new GoString("But when the abstract is known by the Haxe compiler: " + ui8.typeName()));
}

private function printType(a:Any) {
	var typeName = "unknown";
	var f = Reflect.field(a, "typeName");
	if (Reflect.isFunction(f)) {
		typeName = Reflect.callMethod(a, f, []);
	} else if (Std.isOfType(a, Int)) {
		typeName = "int";
	}
	stdgo.fmt.Fmt.println(new GoString(typeName));
}
```
Generates the output:
```
int
Cl
int
But when the abstract is known by the Haxe compiler: uint8
```

There is a similar issue for pointers. Passing a `Pointer` through Haxe `Any`, how will the simulated Go reflection system know that it is a `*int`?

Maybe creating a go2hx `Interface` type for Go `interface{}` containing both the data item and the typeName would be enough to cope with abstract types and pointers etc..

**Comment by PXshadow** [comment]
Oh yes I see your point now.

> Notice from the code above that I think that _typeName_ should contain the Go type name.


So meaning you directly pull from the Go ident, and have for instance "*int" as an end check result?

> The pointers to class examples work using the _is_pointer_ idea, but I'm not yet convinced that pointers to abstract types will work correctly.

I think it should because all abstract classes will be basicTypes and therefore will use the real Pointer abstract, the underlying type for it is PointerData witch can be "is" checked to see if it's a pointer and from there can easily access the underlying type with a .get() function call.

> Personally I think an actual GoBasicTypeInterface<T> class would be more elegant, and easier for users to understand, but it is a matter of taste.

Sure I mean I'd be interested in doing it with a GoBasicTypeInterface type... Actually thinking about this what about a reversed GoBasicTypeInterface abstract where this logic of grabbing if it is a pointer or not, as well as the typeName is handled within a @:from function so it would be extremly elegant as the end generation could actually just be a switch statement of a typeName checking strings.

**Comment by elliott5** [comment]
I agree, "it would be extremely elegant [if] the end generation could actually just be a switch statement of a typeName checking strings" of Go type names.

if a pointer .get() call can return a value that is an abstract class, that would be very helpful - but I don't have time to check.

**Comment by elliott5** [comment]
Certainly extending your example to first point at, then recover, the abstract type does not work.
```
  var pw = new PointerWrapper<GoUInt>(ui);
  var ptr3 = Go.pointer(pw);
  trace(ptr3._value_,Type.typeof(ptr3._value_)); // output: 2,TInt
```
Haxe abstract types only exist at compile-time, not at run-time.

**Comment by PXshadow** [comment]
I wrote out a bunch of methods to build the typeName system, one thing I have no idea how to do is the underlying type of a Slice and Go Array to turn this ``Slice<Int>`` into ``[]int`` I added the interface type to all generated classes based on your reccomendation and it does help, but I'm not sure how to pull the typeName out of it. I could potentially generate somehow a generic for all combinations of a GoArray/Slice/Map and then check its type via the return name, but that seems quite unstable. So bascially how do you get the underlying type with reflection as Go manages to do in Haxe, and I apoligze if I'm missing something fundemental, I'm seeing more and more your point initially of just needing full scale interface generation for each Class. It seems the more I learn the closer I always seem to get to where you advised intially and I just don't have enough experince to figure out all the restrictions that you considered, until after I worked through it. I apperciate the patience as I continue to learn better methologidies to work through the problems, that you are extremly good at figuring out solutions for @elliott5 .

**Comment by PXshadow** [comment]
Slight update I think grabbing the underlying type is doable, it will take an additional system to parse and correctly turn into, but ``Type.getClassName`` as seen here with generic can get the underlying type name as a string. https://try.haxe.org/#90BcdeAe

**Comment by elliott5** [comment]
Obviously, I'm also considering how to implement Go's `reflect` package in the context of go2hx. 
Both Go and TinyGo have a central data-structure containing all the type information for a particular executable (which TinyGo indexes with a bit-encoded type-code held as a `uintptr` and Go indexes with a pointer).
go2hx can't have a central registry of types, because the code is not intended to run alone, but as part of a user's normal Haxe program.

Luckily, Haxe already has the concept of classes (with reflection mechanisms), which are very similar to Go types ... go2hx needs to build on that basis. What is the Haxe equivalent of a Go type switch @PXshadow ?

**Comment by PXshadow** [comment]
@elliott5 there is no real equivlant, the closest would be "is" if chaining, Haxe programs seem to hardly use reflect, generics, macros, and some compiler metadata seems to cover for it.

**Comment by elliott5** [comment]
Yes, the Haxe reflection regime is not as well developed as that for Go. But I suspect go2hx is more likely to succeed by building on what Haxe already has. I'll do more research to see where the gaps are, if any.

**Comment by elliott5** [comment]
It must be possible to call a Haxe function from Go code @PXshadow .
How do you envisage that happening?

**Comment by elliott5** [comment]
In GopherJS they have a specific `js` pseudo-Go module, see https://pkg.go.dev/github.com/gopherjs/gopherjs/js

TardisGo also had a similar `hx` pseudo-Go module, see https://github.com/tardisgo/tardisgo/blob/e0dd9a7e46b597cdc4d310c8a677c2da7444008e/haxe/hx/hx.go

So go2hx need a similar pseudo-Go module. This will allow much of the `reflect` (and maybe other low-level) modules to be written in Go, and therefore able to reuse existing Go code from elsewhere.

**Comment by elliott5** [comment]
I'm currently considering if Go `reflect.Type` could be modelled as an Haxe `enum` built around Go `reflect.Kind` similar in conception to the TinyGo `typecode` idea.

**Comment by PXshadow** [comment]
> It must be possible to call a Haxe function from Go code @PXshadow .
> How do you envisage that happening?

I looked at your hx.go I like the idea of gohaxelib approach you wrote about wanting to do in the future for Tardisgo. When the go2hx library gets imported there could be a flag to enable Go externs so when the Haxe code gets compiled, a global macro could run and generate Go extern code with the Haxe types. I just tested it right now and it can be grabbed that way as long as dead code elimination is not enabled for the extern generating step.

I could also produce something similar to stdgo.Go but the reverse on the Go side for Haxe, not sure what to add for functions there yet, and I'm hoping The abstract system makes it easy to interchanably use go2hx types and native Haxe types.

> I'm currently considering if Go reflect.Type could be modelled as an Haxe enum built around Go reflect.Kind similar in conception to the TinyGo typecode idea.

This is a great idea and in general would be useful to have a system like that for IOTA, I think I've seen something that achieves the requirments for a Go styled enum, an enum abstract using @:from if I remeber correctly.

**Comment by elliott5** [comment]
I've raised #25 to cover the `iota` issue, I hope the `@:from` Haxe magic you speak of will help.

The idea of allowing Go externs to be published in a way that makes the go2hx'd Go library modules more usable is very attractive.

Yes, having Go small-integer types as abstracts over `Int` has the great advantage that calls from regular Haxe code can just use the `Int` type (`UInt` is only an actual type on Flash & C# targets) ...
...but the disadvantage that Haxe runtime reflection cannot detect the abstracts, as they only exist at compile time.

So to give the reflect code a hint, we need to make extra information available.

Where a small integer abstract is passed to an `interface{}/Any` we have already discussed putting a type description into the passed value.

For pointers, we will probably need another access function that returns the value as an `interface{}/Any` to allow them to be typed, eg `*uint8`.

Structures are more problematic, because they don't have accessor functions pre-built:
```
		var data = {uival: cast(42,UInt), bval: true};
```
I've thought of adding the type to the field name (`uival$UInt` above) but that would be unusably ugly very quickly.
So this remains an open issue that will almost certainly require more code to be generated. 
Maybe a structure should be generated as an `interface{}/Any` so that the type info is pre-generated.

Good news on the use of a Haxe `enum` for Go types, it looks feasible so far.

**Comment by elliott5** [comment]
A further complication for Go structs is that each field can have a string `structTag` visible via reflection.
I don't think go2hx currently has a way to capture these strings, which form part of the type information.

See the example here https://golang.org/pkg/reflect/#StructTag .

**Comment by PXshadow** [comment]
Yeah good point, I'm able to pull the tags from the AST but yeah it needs to be possible to store it for reflection package use, perhaps a hidden added variable field to a structure and then Go can attempt to access using a hasField and then getField the hidden variable (hidden being multiple underscores and hard to access or type normally)

**Comment by PXshadow** [comment]
> ...but the disadvantage that Haxe runtime reflection cannot detect the abstracts, as they only exist at compile time.

I think this is only problomatic if it's possible for Go to be able to generate typed numbers without being typed, witch I really hope is not possible because my brain might explode at that point 😅 . As for the pointer, the PointerWrapper PointerType can have the __is__pointer boolean set for structs and standard Pointers for basicTypes can detect that it's a pointer because the underlying abstract type is PointerData. I think the two extra fields being added for each struct not is not to bad, (_is_pointer_, _typeName_) I would much prefer we get things working with solid methdology of course for the descions and then we can go back later and refine, let me know what you think.

**Comment by elliott5** [comment]
I think the idea of using "hidden" fields in Haxe anonymous structures or classes (static fields) to hold Go type information is the right way to go.

It is important that regular Haxe can use the go2hx code as easily as possible, so the go2hx type system needs to build on the Haxe system. Regular Haxe anonymous structures should be treated as Go `struct` types, Haxe classes as Go named types on a Go `struct`.

So, for example, a Haxe class should be able to implement the Go [`fmt.Stringer` interface](https://golang.org/pkg/fmt/#Stringer) simply by having a Haxe function equivalent to a go2hx Go `String() string` method.

I'd hoped that a go2hx type switch could be a simple Haxe Enum switch, and it mostly could be, except for Go interface type matching, which requires comparison of type methods. I'm still working this issue through.

**Comment by elliott5** [comment]
I see that the `AnyInterface` has a field `_typeName_` holding a Go type string. 

For fields in anonymous structures or classes that don't directly map to Go types should there be a hidden field `fieldName_typeName_`: 
```
/**
    T struct {
        I  int
    	U8 uint8
    }
**/
@:access(T_) @:structInit @:allow(command_line_arguments) final class T_ {
    public var i : Int = 0;
    public var u8 : UInt8 = 0;
    public static var u8_typeName_ : String = "uint8";   // Go field type information for abstract type
    //...
}
```

The Go type for the struct could then be constructed by inspecting the fields of the struct.

**Comment by elliott5** [comment]
A type-switch could certainly be created using the `_typeName_` values, composed as Go textual descriptions of the types concerned. 

Some issues would be:
- ensuring that the formatting of the Go type text was always consistent
- maybe needing to use the fully qualified paths to Go named types (see below)
- as mentioned above, string comparison is not enough for Go's ["Duck Typing"](https://medium.com/@matryer/golang-advent-calendar-day-one-duck-typing-a513aaed544d) which requires the inspection of methods/class functions to discover if a named type satisfies an interface. 

As a result, I have concluded that Enums are not necessary to describe Go types.

Named type uniqueness: if two packages in a code set were called "cookiejar" how would the type system tell the difference between them? (There are often two called "log" for example.)
```
package main

import (
	"fmt"
	"net/http/cookiejar"
	"reflect"
)

func main() {
	var j cookiejar.Jar
	jT := reflect.TypeOf(j)
	p := jT.PkgPath()
	fmt.Printf("type is %T but full path is %s\n", j, p)
	// type is cookiejar.Jar but full path is net/http/cookiejar
}
```
So in this case I suggest it would be better for the named type to be called `net/http/cookiejar.Jar` in go2hx. 
Haxe native classes would have fully qualified names with dots in the path.

**Comment by PXshadow** [comment]
> as mentioned above, string comparison is not enough for Go's "Duck Typing" which requires the inspection of methods/class functions to discover if a named type satisfies an interface.

Interface inspection at compile time is implemented. 

I agree with having full path names that should prevent any conflicts as you have mentioned. The only issue I see is Interfaces as you've said, there needs to be some way to not just pass one typeName but rather a list I think to cover Interfaces. If you have any ideas for implementation I'm all ears @elliott5 .

**Comment by elliott5** [comment]
I'll work up some ideas... ;)

**Comment by elliott5** [comment]
To get Go's "Duck Typing" properly requires a working Go `reflect` package that provides the `Implements` method on a type given an `interface`.

go2hx does not have this yet, and it will take quite some time to provide, see the low-level Go code for  `implements`:
https://github.com/golang/go/blob/04e1176fd288f1ceba987d8d2fd9040e45157b38/src/reflect/type.go#L1476

TinyGo does provides this method, but it always returns `false` when testing an `interface` - see:
https://github.com/tinygo-org/tinygo/blob/138befd8a25d2a5c65e2e740038a8c3a44e95222/src/reflect/type.go#L415

I suggest go2hx does the same in the short term, there probably needs to be a separate issue for the `reflect` package (two stages TinyGo emulation, GopherJS/TardisGo emulation).

This means that go2hx cannot generate the correct code for a type switch interface test branch, instead it would have to always be ignored whist the `reflect` package emulates TinyGo. 

Since the `fmt` package relies on this type switch `Implements` functionality, it will not be possible to get it fully working until  phase 2 of the `reflect` package.

**Comment by elliott5** [comment]
Go `type.Implements(interface)` returns `true` if the `type` has all of  the method names, with correct function signatures (correct parameters in/out), required by the `interface`.

Unfortunately Haxe reflection only allows class function names (Go method names) to be found at runtime, not their function signatures.

This will require a work-around to hold the function signature in machine readable form, probably in the form of a `_go_` style static variable in the class.

Native Haxe classes wanting to use Go libraries that did not provide this variable would need special processing. For example Go `String()` method calls could be mapped to Haxe `toString()` function calls - which would make `fmt` work properly for native Haxe.

**Comment by PXshadow** [comment]
> Native Haxe classes wanting to use Go libraries that did not provide this variable would need special processing. For example Go String() method calls could be mapped to Haxe toString() function calls - which would make fmt work properly for native Haxe.

Yeah this makes sense and is doable. As for Haxe native code working with go2hx's std, helper macro functions can be created to add the necessary information to let the std handle them.

> This will require a work-around to hold the function signature in machine readable form, probably in the form of a _go_ style static variable in the class.

An example of this would be very helpful, I think I understand the problem that being, at runtime Haxe's reflect has no means to inspect an interfaces methods because it's not a runtime value able to be inspected by reflect in Haxe, so signatures would be the data of the interface methods to be able to compare against somehow on runtime, or have a list of types that implement the interface.

> Since the fmt package relies on this type switch Implements functionality, it will not be possible to get it fully working until phase 2 of the reflect package.

I really liked the phase plan, however not having fmt until phase 2 is a pretty big limiter. Also go2hx is now ahead of Tinygo when it comes to recover, almost entirely because you built a fantastic conception to complete the recover system. Perhaps we can craft a solid system and be able to implement this as well and start making major progress into reflect, fmt, and typeswitches. I can take a shot at designing a working typeswitch interface tommrow.

If skipping to stage 2 doesn't make sense for this, I will just continue with stage 1 reflect. Let me know @elliott5 what you think the best course of action is.

**Comment by elliott5** [comment]
I'm very flattered @PXshadow, thank you for your kind words, I can see you're keen to press ahead of TinyGo, which is very laudable, but will take some time. I'll work up a `reflect` prototype over the next week or so.

The type-switch needs to use the same data-structures as are used by `reflect`, so please don't do any more work on it until you see my prototype.

**Comment by PXshadow** [comment]
Of course 😄, and yes I want to catch up as soon as possible with supporting the Go spec. I am really looking forward to you desinging the reflect prototype! I won't work on reflect until I see your prototype and instead working on supporting other parts of the spec not releated using mostly tinygo's tests.

**Comment by elliott5** [comment]
Each `case` in a type switch will likely need to have a type created to test against.

I have been considering how best to extend the "_typeName_" idea to allow constant type descriptions.

@PXshadow, I assume somewhere in the parse tree is a branch containing these declarations that could be used to generate the type descriptions.

What is the best way to go about doing this?

**Comment by elliott5** [comment]
FYI @PXshadow , I've refactored GT_struct to make it easier for go2hx to generate declarations, but not committed it yet.

**Comment by PXshadow** [comment]
Feel free to commit it, my recommendation for changing the type switch behavior of go2hx (as asked about above) would be making a new branch and modifying the [typeSwitchStmt function](https://github.com/go2hx/go2hx/blob/master/src/Typer.hx#L873).

**Comment by elliott5** [comment]
see https://github.com/go2hx/go2hx/issues/24#issuecomment-829047289

I'm not sure a new branch is required, you are the best judge of that.

**Comment by PXshadow** [comment]
@elliott5 yeah you're right no new branch needed, was thinking of another set of potential changes.

**Comment by PXshadow** [comment]
Should now be functional, there is a slight deviation away from the spec for int as an alias as well for int32/int64 not sure if that's a good idea or not but currently it's working. Any issues with an error case for typeswitching would be very helpful.
